### DISCUSSION TITLE
Redefinition of namespaces

### DISCUSSION THREAD

**Discussion by LeftTwixWand** [dotnet/csharplang#5490] [Language Ideas] [answered]
Hi. I'm working on a complex open source library, that would help users to start working on WinUI 3 applications with DDD project organization. Because there are some solutions for web projects, but they won't work for desktop. It's a real problem to find something about DDD for desktop, especially for such new technology, like WinUI 3.

That's why I want to create my own library and Visual Studio Templates.
And now it's time to add some abstractions for Repository patter and Specification pattern.
I can write code by myself, I can copy the source code from somewhere else. But If there is already exist an excellent NuGet package with the functionality, that I'm looking for - why not to use it?

But here the problem begins. Different namespaces will confuse newcomers.
That's why I would like to do something like this in my library:
```csharp
global using DDDBlocks.Domain.Specifications = Ardalis.Specification.
```

I want to override / redefine namespace of other NuGet library.
Now, it's possible to use alias usings, but it works only inside one file. I want to see it for the entire project:
```csharp
using GenericCollections = System.Collections.Generic;
```

I think, I will be a good step for open source community, because we could combine small libraries into complex solutions, like project templates. That will definitely increase speed of development.

**Comment by HaloFour** [comment]
Global aliases shipped in C# 10.  That said, I would be pretty annoyed if a project template added a global alias to make it seem like some other library existing under their namespaces.

**Reply by LeftTwixWand** [reply]
Maybe you are right, and I will be confusing, when someone will try to find the source code of some feature, that is not part of my library.

**Comment by ufcpp** [comment, marked as answer]
You can already do that. NuGet packages can contain a `props` file like

```xml
<Project>
  <ItemGroup>
    <Using Include="System.Collections.Generic" Alias="GenericCollections " />
  </ItemGroup>
</Project>
```

**Reply by LeftTwixWand** [reply]
Thank You so much!

**Reply by jnm2** [reply]
https://docs.microsoft.com/en-us/nuget/create-packages/creating-a-package#include-msbuild-props-and-targets-in-a-package for more on this.

**Reply by LeftTwixWand** [reply]
Thanks!!
